### ISSUE TITLE
Solution for Linux/Mac CI?

### ISSUE THREAD

**Issue by danluu** [BitFunnel/BitFunnel#124]
With Travis, we have (at least) three serious issues:
1. The vast majority of failures are due to flakiness or changes on their end, making the CI results noisy.
2. When we get failures due to changes on their end, it's often time consuming to fix the failures. Over the past few months, I think Alex has spent a day (or more?) on CI fixes, and I've spent probably two days on CI. This is more disruptive than it sounds because the time is often an hour and an hour there.
3. Travis is slow. We can compile a clean build locally in 10s and run tests in 20ms. On Travis we queue for a couple minutes and then run for 10 minutes. It looks like we could pay more money and speed it up to the point where it "only" takes 2 minutes to finish a CI run, but that's still an order of magnitude slower than Appveyor.

On 1 & 2, if you look at the Travis bug tracker, it's easy to see that [C++ CI is a low priority for them](https://github.com/travis-ci/travis-ci/issues/6300), and they're not really interested in making [changes that would make C++ CI](https://github.com/travis-ci/travis-ci/issues/5959) [easier to set up](https://github.com/travis-ci/travis-ci/issues/5821) for anyone who wants to use a recent compiler.

On 3, Travis is allegedly faster if you have a `travis.yml` that has `sudo: false`, but we can't do that at the moment because we rely on apt-get. It looks like [hana has a .travis.yml that installs modern compilers with without sudo](https://github.com/boostorg/hana/blob/master/.travis.yml), by manually downloading and running the source and install process via what is effectively a giant shell script. From looking at their actual CI results, it doesn't appear that they get much speedup from that doing, although maybe they get more reliability? The author of that yaml file [indicates that he spends a ton of time tweaking it to keep it up to date (with breaking Travis changes???)](https://github.com/travis-ci/travis-ci/issues/6300#issuecomment-232842843), so I don't think that would solve (2).

I've looked into CircleCI a bit, and it looks like it's likely to be substantially faster, perhaps solving (3), but they also appear to not really prioritize C++ support.

On (3), I believe AppVeyor is substantially faster than Travis because we're running on a machine that's configured with a C++ compiler, so we don't have to go and install `clang` and all of its dependencies before our tests start.

If we do nothing, we'll have working (and pretty good) Windows CI, and no Linux or Mac CI. I think that's actually as long as no one wants to contribute to this project, but that we'll have problems pretty quickly if any outside contributors want to work on this. It's understandable that neither Travis nor Circle prioritize C++, but it means that there doesn't appear to be a passable off-the-shelf CI solution for us.

**Comment by danluu** [comment]
[Here's another Travis flakiness issue](https://www.traviscistatus.com/incidents/11hp8bhkrkn7).

**Comment by danluu** [comment]
I tried CircleCI. It asks for "Full control of private repositories" in its permissions. Before granting the permissions, none of the repositories in our org (even the public ones) showed up as possible projects to add. After granting those permissions, it still didn't show any repos in our org (public or private) and I revoked the permissions.

**Comment by danluu** [comment]
Appveyor appears to support clang and has a relatively modern version installed on their VM image (they have 3.8.0 and the latest release is 3.8.1 with 3.9 coming out this month). I wonder if we could use their clang support. The vast majority of the issues on our Linux build aren't Linux issues per se, but issues with clang or gcc. If we could Windows/clang CI, I suspect that would catch 90% of our Linux/clang issues.

**Comment by tn0502** [comment]
I was looking for Travis CI alternatives when I came across your issue and was wondering if you've looked at [drone.io](https://drone.io/)? I'm in a similar situation as yours but the alternatives I've seen are all based on Trusty so you would still have to manually install a more recent compiler, tools, etc. Another alternative is to use Visual Studio Online and set up a box yourself. The downside is that you'll have to set up your repo to also push to VSO.

**Comment by danluu** [comment]
@tn0502 Thanks for the suggestion! We haven't looked into that. Their docs claim that [they support clang (no version listed), gcc 4.6, and gcc 4.8](http://docs.drone.io/cpp.html).

YMMV, but I believe that the limiting factor for us in terms of version right now is that we use `compare_exchange_weak`, [which requires at least gcc 4.8 to fix a known bug](https://github.com/BitFunnel/BitFunnel/issues/35), so we might be able to get away with using drone. If you end up using them, please let us know how it goes.

I think that for us the problem is that we're tentatively planning to use the "new" C++ file/path support, but based on their listed gcc versions, they don't have super modern compilers, which is likely to be a problem.

**Comment by Conan-Kudo** [comment]
@danluu You could try out GitLab CI, which is flexible enough to support any type of language development using any operating system (and flavors of them!).

More info available at: https://about.gitlab.com/gitlab-ci/

**Comment by andoma** [comment]
@danluu saw this issue your tweet yesterday and I've been working on a CI / autobuild system called Doozer which is currently in beta mode. https://doozer.io/  

It's less focused on node/rails/etc but rather for native apps. For example I use it to build complete Linux buildroot systems, etc

It offers Linux build environments from Ubuntu 12.04 up to 16.04 and the user is free to install whatever dependencies he/she likes. The build agent uses btrfs snapshots to retain, recall and reset build environment for faster builds.

I tested to build BitFunnel using it, see https://doozer.io/user/andoma/BitFunnel with the corresponding config here: https://github.com/andoma/BitFunnel/blob/master/.doozer.json

It fails to build with gcc-5 though (-Werror trips)

Anyway, please have a look and, as usual, any kind of feedback would be appreciated.

**Comment by Conan-Kudo** [comment]
@andoma This looks quite interesting. Would you be willing to add CentOS/Fedora to your Linux environments?

**Comment by andoma** [comment]
> @andoma This looks quite interesting. Would you be willing to add CentOS/Fedora to your Linux environments?

That's definitely on the radar so I can prioritize it for sure

**Comment by Conan-Kudo** [comment]
@andoma I'd be interested in using it for my projects on GitHub, GitLab, and BitBucket if it supports CentOS and Fedora.

**Comment by danluu** [comment]
@andoma Thanks! This is amazing! We'll have to try this out. We had some known issues with gcc that I believe @hausdorff has now fixed, so that failure was "expected". When I have a bit of time, I'll set this up and see how it goes!

Thanks again!

**Comment by danluu** [comment]
@andoma I tried adding your .doozer.json and creating an account. Account creation seems to have worked fine, but our builds seem to have been queued for a few hours. Is there a status page where we can see what's going on?

**Comment by andoma** [comment]
@danluu Seems I somehow failed to correctly restart the production build agent when i Added Fedora and Centos environments. Sorry about that :-(

It should be fixed now and I've also made sure it respawns correctly as a system service on the host.

**Comment by danluu** [comment]
No problem! A test now fails on #298, which is a known issue for us. This is great! Thanks again.

**Comment by danluu** [comment]
@andoma I tried pushing a quick change to see what would happen now that the queue is unblocked. The build is really fast! Probably about the same speed as our appveyor build because everything we're using is pre-installed on the machine. If this something you can provide to folks in general who use C++, I think this has a really good chance of catching on.

One question I have is, how can we enable this for PRs (or is it enabled by default)? I skimmed the docs and it wasn't super obvious to me how it works since it seemed like you have to add individual branches (like master).

**Comment by andoma** [comment]
Building of PRs are enabled by default. There really should be an option to turn it off per project but that's also on my TODO list :-)

I just sent an PR that include the test output as an artifact.

**Comment by andoma** [comment]
Btw, you should probably remove the projects webhook for the danluu/BitFunnel-1 doozer project because right now it will trig a build in that project for PRs as well.

Once I add a [Delete Project] button it will of course clean up those things automatically. 

I assume you added BitFunnel-1 by mistake or so.

**Comment by danluu** [comment]
Correct -- I clicked the wrong thing, got that project created, and couldn't figure out how to remove it 😄 . Thanks, I'll make that change.

**Comment by danluu** [comment]
@andoma Is there a way to configure who gets emailed on build failures? I skimmed the docs and didn't see anything.

**Comment by andoma** [comment]
@danluu no email support atm. But notifications is also on the prio list (obviously). Email would be some kind of minimum level I suppose. IRC, Slack and Telegram is also on my personal wishlist.

I just need to find some decent email sending service. I expect that if I just send things using sendmail myself it would likely end up in spam filters.

But there is a [Delete project...] button in the project settings page now so you can get rid of BitFunnel-1 :)

**Comment by andoma** [comment]
Guys, Just want to let you know I'm aware of the pesky "No output detected for 600 seconds" failing builds. It seems that the build agent hangs inside a fork() call when it's preparing to execute a command. The agent has been doing tens of thousands of builds for my personal projects before I made Doozer a public service and I've never seen this before. When making doozer public the build agent has largely remained unchanged but apparently something has happened here.

I'll try to fix it ASAP.

Also I'm gonna add some kind of rescheduling mechanism if a build fails for "external reasons" like network issue, etc. Perhaps "silent output" could be classified to that as well.

**Comment by danluu** [comment]
I wonder if we're doing something that contributes to that? We recently had an appveyor flaky build failure with similar-ish symptoms:

https://ci.appveyor.com/project/MichaelHopcroft/nativejit/build/1.0.89

Re-running the same build twice gave us two passing runs instead of a hang:
https://ci.appveyor.com/project/danluu/nativejit/build/1.0.30
https://ci.appveyor.com/project/danluu/nativejit/build/1.0.31

This could just be a coincidence, though. The lack of useful logging when that happened makes it tough to figure out what actually went wrong.

**Comment by danluu** [comment]
@andoma Just in case you haven't seen this, we're now seeing cases where the build fails, seemingly because doozer can't connect to github. The failure message is:

~~~
Status: GIT: Unable to connect to https://github.com/BitFunnel/BitFunnel.git -- Failed to connect to github.com: Connection timed out
~~~

**Comment by andoma** [comment]
@danluu yes i saw this now as well. I logged onto the agent and couldn't access github via curl and did some additional network troubleshooting (traceroute, and checked using looking glasses at the ISP) and it seems it was a connectivity issue.

It seems better now though. Btw, you can right-click on a build to get a context menu and re-run the build.

Anyway, the plan is to move the agent off a fixed machine I have and instead spawn machines on a cloud provider (probably GCE because their minimum charge limit is 10 minutes).

**Comment by andoma** [comment]
> @andoma Is there a way to configure who gets emailed on build failures? I skimmed the docs and didn't see anything.

There is rudimentary email support now. There is a new option in project settings where you can tell Doozer to mail the person doing the git push with build results (can also be set to only mail on failure).
Currently this is only supported on GitHub as the other integrations does not reveal the email address of the pusher in the webhook call.

I'm also working on a more general notification configuration where you can add any number of recipients that will get notifications when something happens.

**Comment by danluu** [comment]
Thanks! That by itself already delivers 90% of the value we'd get out of any sort of notification setup.

**Comment by danluu** [comment]
I'm going to close this issue because doozer looks like it's good enough that we don't have to think about this again unless something about doozer changes or this project grows *a lot*. We've had some flakiness, but not nearly as much as on Travis, and the build is much faster.

We still don't have a solution for Mac CI, but the vast majority of failures on Mac are also reproducible on Linux, so we're not missing a ton of coverage there. I'd like to fix that at some point, but I don't think Mac CI is even in the top 100 priorities for us right now.

Thanks again to @andoma for both building doozer and suggesting it to us.

**Comment by AraHaan** [comment]
@andoma the issue with doozer is when I want to install clang 3.9 on 1 of my projects with it or even when I need to use homebrew to install gcc or clang (and then there would be no way for me to for now install homebrew on it as it requires admin privs). Doozer is better and faster than Travis. However it lacks some things just like travis does. I would like if Doozer actually had homebrew preinstalled on those osx's so the build does not fail because it did not find an command ``brew``. Or if it could not find `clang-3.9` to install.

**Comment by tn0502** [comment]
Homebrew is preinstalled. Docs say so and I've been able to install deps
for my builds.

Here's my configuration:
https://bitbucket.org/tido/rainbow/src/d0670d97437bab97d65db95faaa3419a788d75f3/.doozer.json?at=master&fileviewer=file-view-default

**Comment by AraHaan** [comment]
I see, thanks. Now I think the only thing I have of issue is trusty on clang 3.9.

**Comment by AraHaan** [comment]
Ok, so I have an issue with the OSX ones now. On Hopebrew on the OSX build it seems to fail to find the clang formulae that can be found in the https://github.com/Homebrew/homebrew-core repo. Oddly it does not. For more information see https://github.com/Homebrew/homebrew-core/issues/10657 @andoma
